File: string/string-1/10.py
# Tc = O(m.n)
# Sc = O(m.n)
class Solution:
    def solve(self, i, j, s, p, t):
        if j == len(p):
            return i == len(s)
        if t[i][j] != -1:
            return t[i][j]
        first_char_matched = False
        if i < len(s) and (p[j] == s[i] or p[j] == '.'):
            first_char_matched = True
        if j + 1 < len(p) and p[j + 1] == '*':
            not_take = self.solve(i, j + 2, s, p, t)
            take = first_char_matched and self.solve(i + 1, j, s, p, t)
            t[i][j] = not_take or take
        else:
            t[i][j] = first_char_matched and self.solve(i + 1, j + 1, s, p, t)
        return t[i][j]
    
    def isMatch(self, s: str, p: str) -> bool:
        m, n = len(s), len(p)
        t = [[-1 for _ in range(n + 1)] for _ in range(m + 1)]
        return self.solve(0, 0, s, p, t)



from functools import lru_cache
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        @lru_cache(None)
        def solve(i: int, j: int) -> bool:
            if j == len(p):
                return i == len(s)
            
            first_char_matched = i < len(s) and (p[j] == s[i] or p[j] == '.')
            
            if j + 1 < len(p) and p[j + 1] == '*':
                not_take = solve(i, j + 2)
                take = first_char_matched and solve(i + 1, j)
                return not_take or take
            
            return first_char_matched and solve(i + 1, j + 1)
        
        return solve(0, 0)

File: string/string-1/10.cpp

// Tc = O(2^(max m, n))
// Sc = recursion tree max depth(max length of the string or the pattern) => O(n) or O(m)
class Solution {
public:
    
    bool isMatch(string text, string pattern) {
        if (pattern.length() == 0) {
            return text.length() == 0;
        }

        bool first_char_matched = false;
        if(text.length() > 0 && (pattern[0] == text[0] || pattern[0] == '.')) {
            first_char_matched = true;
        }
        if (pattern.length() >= 2 && pattern[1] == '*') {
            return (isMatch(text, pattern.substr(2)) ||
                    (first_char_matched && isMatch(text.substr(1), pattern)));
        } else {
            return first_char_matched && isMatch(text.substr(1), pattern.substr(1));
        }
    }
};


//Approach-2 (Recursion + Memoization)
//T.C : O(m*n)
//S.C : O(m*n)
class Solution {
public:
    int t[21][21]; //O(m.n)
    bool solve(int i, int j, string s, string p) {
        if (j == p.length()) {
            return i == s.length();
        }  
        if(t[i][j] != -1){
            return t[i][j];
        }  
        bool first_char_matched = false;
        if ( i < s.length() && (p[j] == s[i] || p[j] == '.')) {
            first_char_matched = true;
        }
        if(j+1 < p.length() && p[j+1] == '*'){
            bool not_take = solve(i, j + 2, s, p);
            bool take = first_char_matched && solve(i + 1, j, s, p);
            return t[i][j] = not_take || take;
        }
        return t[i][j] = first_char_matched && solve(i + 1, j + 1, s, p);
    }
    
    bool isMatch(string s, string p) {
        memset(t, -1, sizeof(t));
        return solve(0, 0, s, p);
    }
};